spec/lang/intrinsic.md

```diff
@@ -194,21 +194,33 @@ impl<M: Memory> Machine<M> {
         };
 
         let func = self.fn_from_addr(ptr.addr)?;
+        if func.args.len() != 1 {
+            throw_ub!("invalid first argument to `Intrinsic::Spawn`, function should take one argument.");
+        }
 
-        if func.args.len() != 0 {
-            throw_ub!("invalid first argument to `Intrinsic::Spawn`, function takes arguments");
+        let (data_ptr, data_ptr_ty) = arguments[1];
+        if !matches!(data_ptr_ty, Type::Ptr(PtrType::Raw)) {
+            throw_ub!("invalid second argument to `Intrinsic::Spawn`, data pointer should be a raw pointer.");
+        }
+        // This logic is taken from `eval(Terminator::Call)` and `eval_argument`.
+        let data_ptr_pty = PlaceType::new(data_ptr_ty, M::T::PTR_ALIGN);
+        if !Self::check_abi_compatibility(data_ptr_pty, func.locals[func.args[0]]) {
```

Comment:
This check should be in `spawn`. It's not great to have the check in one place and the logic that relies on the check in a completely different file.